lib/apiservers/engine/backends/volume.go

```diff
@@ -96,12 +105,103 @@ func (v *Volume) Volumes(filter string) ([]*types.Volume, []string, error) {
 		log.Infof("%s", vol.Name)
 		volumeMetadata, err := extractDockerMetadata(vol.Metadata)
 		if err != nil {
-			return nil, nil, fmt.Errorf("error unmarshalling docker metadata: %s", err)
+			return nil, nil, derr.NewErrorWithStatusCode(fmt.Errorf("error unmarshalling docker metadata: %s", err), http.StatusInternalServerError)
 		}
 		volume := NewVolumeModel(vol, volumeMetadata.Labels)
 		volumes = append(volumes, volume)
 	}
-	return volumes, nil, nil
+
+	// Parse and validate filter args
+	volumeFilters, err := filters.FromParam(filter)
+	if err != nil {
+		return nil, nil, derr.NewErrorWithStatusCode(err, http.StatusInternalServerError)
+	}
+	if err := volumeFilters.Validate(acceptedVolumeFilters); err != nil {
+		return nil, nil, derr.NewErrorWithStatusCode(err, http.StatusInternalServerError)
+	}
+
+	if volumeFilters.Len() == 0 {
+		return volumes, nil, nil
+	}
+
+	// If the dangling filter is specified, gather required items beforehand
+	dangling := false
+	joinedVolumes := make(map[string]struct{})
+	if volumeFilters.Include("dangling") {
+
+		// Validate dangling filter's usage (per Docker code)
+		if volumeFilters.ExactMatch("dangling", "true") || volumeFilters.ExactMatch("dangling", "1") {
+			dangling = true
+		} else if !volumeFilters.ExactMatch("dangling", "false") && !volumeFilters.ExactMatch("dangling", "0") {
+			return nil, nil, fmt.Errorf("invalid filter 'dangling=%s'", volumeFilters.Get("dangling"))
```

Comment:
`dangling=false` is valid. This is checking for supported `dangling` filter formats: `true`, `false`, `1` and `0` (per Docker).